src/sinol_make/commands/gen/gen_util.py

```diff
@@ -0,0 +1,135 @@
+                    will return any ingen matching "*ingen.*"
+    :param ingen_path: path to ingen source file
+    :return: path to ingen source file or None if not found
+    """
+
+    if ingen_path is not None:
+        if os.path.exists(ingen_path):
+            return ingen_path
+        else:
+            util.exit_with_error(f'Ingen source file {ingen_path} does not exist.')
+
+    if task_id is None:
+        task_id = '*'
+    ingen = glob.glob(os.path.join(os.getcwd(), 'prog', task_id + 'ingen.*'))
+    if len(ingen) == 0:
+        util.exit_with_error(f'Ingen source file for task {task_id} does not exist.')
+
+    # Sio2 first chooses shell scripts, then non-shell source codes.
+    if os.path.splitext(ingen[0])[1] == '.sh' and len(ingen) > 1:
+        return ingen[1]
+    else:
+        return ingen[0]
+
+
+def compile_ingen(ingen_path: str, args: argparse.Namespace, weak_compilation_flags=False):
+    """
+    Compiles ingen and returns path to compiled executable.
+    """
+    compilers = compiler.verify_compilers(args, [ingen_path])
+    ingen_exe, compile_log_path = compile.compile_file(ingen_path, package_util.get_executable(ingen_path), compilers, weak_compilation_flags)
+
+    if ingen_exe is None:
+        util.exit_with_error('Failed ingen compilation.', lambda: compile.print_compile_log(compile_log_path))
+    else:
+        print(util.info('Successfully compiled ingen.'))
+    return ingen_exe
+
+
+def get_correct_solution(task_id):
+    """
+    Returns path to correct solution for given task.
+    :param task_id: task id, for example abc
+    :return: path to correct solution or None if not found
+    """
+    correct_solution = glob.glob(os.path.join(os.getcwd(), 'prog', task_id + '.*'))
+    if len(correct_solution) == 0:
+        util.exit_with_error(f'Correct solution for task {task_id} does not exist.')
+    return correct_solution[0]
+
+
+def compile_correct_solution(solution_path: str, args: argparse.Namespace, weak_compilation_flags=False):
+    """
+    Compiles correct solution and returns path to compiled executable.
+    """
+    compilers = compiler.verify_compilers(args, [solution_path])
+    correct_solution_exe, compile_log_path = compile.compile_file(solution_path, package_util.get_executable(solution_path), compilers,
+                                weak_compilation_flags)
+    if correct_solution_exe is None:
+        util.exit_with_error('Failed compilation of correct solution.',
+                                  lambda: compile.print_compile_log(compile_log_path))
+    else:
+        print(util.info('Successfully compiled correct solution.'))
+
+    return correct_solution_exe
+
+
+
+def run_ingen(ingen_exe):
+    """
+    Runs ingen and generates all input files.
+    :param ingen_exe: path to ingen executable
+    :return: True if ingen was successful, False otherwise
+    """
+    is_shell = os.path.splitext(ingen_exe)[1] == '.sh'
+    if is_shell:
+        os.system(f'chmod +x {ingen_exe}')
```

Comment:
Does it work correctly on other operating system? Does e.g. Windows have a `chmod +x` command? Reminder that some operating systems can have different shell commands.